doc/cookbook/translation-bot.md

```diff
@@ -0,0 +1,417 @@
+* Create a Telegram bot by following [this document](https://core.telegram.org/bots#3-how-do-i-create-a-bot), set its name (EaseTranslateBot is used in this article), write down its token, and [set up a Webhook](https://core.telegram.org/bots/api#setwebhook) that points to the Easegress instance installed in the previous step. Our bot will receive notifications of new messages through this Webhook.
+* AWS Access Key ID and Access Key Secret, and ensure that you can use the AWS translation API with this Access Key.
+* Google Cloud's Token and ensure that you can use Google Cloud's Speech Recognize API and OCR (Image Annotation) API with this Token.
+
+It is fine for you to use other vendors' translation, speech recognition, or OCR APIs, but this will require you to adapt the examples in the later sections accordingly.
+
+## 2. How It Works
+
+The sequence chart below shows the workflow of this bot:
+
+![sequence chart](../imgs/translation-bot-sequence-chart.png)
+
+Upon receiving a notification of a new message from the Telegram server via webhook, the bot first checks the message type and does the following accordingly：
+
+* **Text Message**: Extract message text directly；
+* **Voice Message**: In this case, the message body only contains the ID of the voice file, so we need to call Telegram's API to convert the ID to a file address, then download the file and send its contents to Google's voice recognition service to convert it to text；
+* **Photo Message**: This case is basically the same as the voice message, but the file content is sent to Google's Image Annotation service.
+
+After the above processing, all three types of messages are turned into text, and then AWS translation service can be called to translate them into target languages, the target languages used in this example are Chinese, Japanese, and English.
+
+## 3. Pipeline
+
+First, let's check the overall flow that Pipeline orchestrates:
+
+```yaml
+flow:
+# we put the final response builder at the top because Telegram requires
+# us to return a response for every Request, but we only process some of
+# the requests. If we put it at the end, the requests we don't process
+# will end the process early and no Response will be returned.
+- filter: buildFinalResponse
+ 
+# check message type and jump accordingly
+- filter: checkRequestType
+  jumpIf:
+    result0: processText             # text
+    result1: processVoice            # voice
+    result2: processPhoto            # photo
+    "": END                          # ignore, end the processing
+ 
+# text message
+- filter: requestBuilderExtractText
+  alias: processText                 # alias of the filter
+  namespace: extract                 # namespace the filter belongs to
+  jumpIf:                            # conditional jump, begin translation
+    "": translateChinese             # if everything is fine, or end the
```

Comment:
why do use the empty key?

Reply:
because it needs to jump if everything is fine, to skip `processVoice` and `processPhoto`.